### ISSUE TITLE
Adding XML functionality for XMAS!!!!

### ISSUE THREAD

**Issue by dpoetry** [ruffle-rs/ruffle#2007]
I am having a problem with pulling in an XML file on top of the SWF file. This might seem overly complicated. But I'll try to explain. The html file loads p1.swf. Then p1.swf uses LoadMovie to load a p1add.swf file. Then the p1add.swf file uses xml to load a bunch of images/jpgs. So either the p1add.swf isn't loading properly or the XML file inside the p1add.swf file isn't loading.

I am guessing the issue is it not loading the XML file. As other sections of this work use the LoadMovie function just fine and load external SWFs on top of other SWFs.

Here is the overall work: http://www.dpoetry.com/ruffletest/wittenoom2/starthere.html

And here is the page in with the XML files not loading: http://www.dpoetry.com/ruffletest/wittenoom2/artwork/p1.html

And here are the relevant files:

The HTML file: http://www.dpoetry.com/ruffletest/wittenoom2/artwork/p1.html
The primary SWF file: http://www.dpoetry.com/ruffletest/wittenoom2/artwork/p1.swf
The LoadMovie SWF file: http://www.dpoetry.com/ruffletest/wittenoom2/artwork/p1add.swf
The XML File: http://www.dpoetry.com/ruffletest/wittenoom2/artwork/images.xml
All the files: http://www.dpoetry.com/ruffletest/wittenoom2/artwork/

LASTLY: This is what the page above should look and function like: http://www.secrettechnology.com/wittenoom/artwork/p1.html. (will have to enable flash)

NOTE: The above URL from secrettechnology does not have Ruffles loaded. So everything still works within Flash.

Does any of that make sense?

cheers, Jason Nelson

**Comment by Herschel** [comment]
Haven't investigated thoroughly yet, but we should support XML, but don't currently support loading external images.

**Comment by jrob007** [comment]
I have a separate XML file that gets loaded and parsed into a frame as well. It relies primarily on `XML.load()`, `XML.onLoad()` and `XML.parseXML()`. However according to https://github.com/ruffle-rs/ruffle/issues/286 only the `parseXML` method is currently supported by Ruffle. My animation loads, but the textbox simply says "undefined" as it's not loading any external XML file.

**Comment by Herschel** [comment]
The tracking issue is out of date; we should `XML.load/onLoad/parseXML`, so the issue is possibly elsewhere. If possible, could you post a sample SWF? Thank you!

**Comment by dpoetry** [comment]
Sorry for the delay!  I have two swf-based artworks that both use an external XML to load external images.   The first is an artwork I created called Sydney's Siberia and it's pretty complex in terms of how it maps the images and then creates an infinitely zooming mosaic. 

You can download the entire work here: http://www.dpoetry.com/ruffletest/zips/siberia.zip

The second artwork is part of a larger work called Wittenoom.  It also loads external images via xml, with the images falling on the screen, each clickable to be read. 

http://www.dpoetry.com/ruffletest/zips/open-p1swf.zip

And just to share.....I created a range of art games back in the heyday of Flash...including "game, game, game and again game" and "I made this. you play this. we are enemies".  And I'm testing them later today and will load the results. 

But....many, many thanks for helping to keep some of these older works alive. 

cheers, Jason Nelson

**Comment by jrob007** [comment]
> 
> 
> The tracking issue was out of date, I went ahead and updated it; we should support `XML.load/onLoad/parseXML`, so the issue is possibly elsewhere. If possible, could you post a sample SWF? Thank you!

Sorry for the delayed response (holiday and all). I went back and looked at the SWF code and realized it may actually be the `ignoreWhite = true` property that it's fumbling on and which, according to the updated XML tracking, Ruffle doesn't currently support? At least I can try a bit more tinkering in the mean time. The actual code for loading the external SWF is:

```
_root.control.TextBox.html = true;
_root.control.TextBox.wordWrap = true; 
_root.control.TextBox.multiline = true;

//The path to the XML narrative relative to the SWF
var xmlFile = "B01.xml";
var arr = new Array();
narrative = new XML();
narrative.ignoreWhite = true;
narrative.load(xmlFile);
narrative.onLoad = function(success) {
	if (success) {
		frameCount = narrative.firstChild.childNodes.length;
		var xmlObj:XML = new XML();
		
		xmlObj.parseXML(narrative);
		xmlNode = narrative.firstChild.firstChild;
		for (i=0; i<frameCount; i++) {
			arr[i] = new Array();
			for (j=0; j<xmlNode.childNodes.length; j++) {
				arr[i][j] = xmlNode.childNodes[j];
			}
			xmlNode = xmlNode.nextSibling;
		}
		_root.control.TextBox.text = arr[1][0];
	}
}
```
Each sequential frame in the SWF would have `_root.control.TextBox.text = arr[x][0];` where `x` is the frame number so it loads the next node into the TextBox. Anyway, here's the unmodified [Ruffle Test version running on our server](https://www.shipcopumps.com/tools/virtualpump/files/index3.php) and here's the [SWF/XML files](https://www.dropbox.com/sh/v3m3ngs78m9fyi9/AACiJvaPLOzCqBjRPRpdh9fAa?dl=0) if you want to take a look. I really love the work you all are doing! Ruffle is pretty awesome.

**Comment by Herschel** [comment]
@jrob007 
I implemented `ignoreWhite` this past week, so looks like it is working now if you try on the latest nightly build. I've updated the tracking issue.

**Comment by jrob007** [comment]
> 
> 
> @jrob007
> I implemented `ignoreWhite` this past week, so looks like it is working now if you try on the latest nightly build. I've updated the tracking issue.

Awesome! I uploaded the latest Ruffle build and it is working! :-) The only minor issue now that I might have to fix on my own is removing the XSL transform and the title headers from each of the XML files (I'm not even sure why I did that back then, but it worked at the time). Anyway, that may be an entirely different issue with regards to Ruffle, but at least the data from XML is loading within the textbox for each frame and is entirely user-readable. That's far more than I could hope for with these SWF animations. I really appreciate the work you guys are doing with Ruffle, it's phenomenal.
